neo/Network/P2P/Payloads/WitnessScopeType.cs

```diff
@@ -8,20 +8,31 @@ public enum WitnessScopeType : byte
         Global = 0x00,
 
         /// <summary>
-        /// RootAccess means that this condition must hold: EntryScriptHash == CallingScriptHash
-        /// No params is needed, as the witness/permission/signature given on root of invocation will automatically expire if entering deeper internal invokes
+        /// EntryOnly means that this condition must hold: EntryScriptHash == CallingScriptHash
+        /// No params is needed, as the witness/permission/signature given on first invocation will automatically expire if entering deeper internal invokes
         /// This can be default safe choice for native NEO/GAS (previously used on Neo 2 as "attach" mode)
         /// </summary>
-        RootAccess = 0x01,
+        EntryOnly = 0x01,
 
         /// <summary>
         /// Custom hash for contract-specific
         /// </summary>
         CustomScriptHash = 0x02,
 
+        // 0x03 -> composition between EntryOnly and CustomScriptHash. 
+        // Example: we invoke NEO native transfer using 0x01 (it should work).
+        // We invoke NEO native transfer using 0x03, and custom hash for GAS asset (it should fail)
+
         /// <summary>
         ///  Custom pubkey for group members
         /// </summary>
-        ExecutingGroupPubKey = 0x03
+        ExecutingGroupPubKey = 0x04
+
+        // 0x05 -> composition of Group + EntryOnly. See example for 0x03.
+
+        // 0x06 -> composition of Group + Custom. Probably, Group hash should be considered instead of CustomHash.
+        // 0x06 should be equivalent to 0x04
+        
+        // 0x07 -> this should be equivalent to 0x05 (since 0x06 -> 0x04) + 0x01
```

Comment:
@erikzhang @shargon what do you think of these enum compositions? reasonable?

Reply:
Compositions work as an intersection here.
Global \inter EntryOnly == EntryOnly
Global \inter CustomHash == CustomHash
(means that Global won't affect anything on compositions)
EntryOnly \inter CustomHash == EntryOnly && CustomHash
etc

Reply:
Maybe the compositions are confusing, I prefer simpler types

Reply:
If we allow multiple CosignerUsage on header for same ScriptHash, which may be needed for two custom hashes simultaneously (without becoming global), then this will become unnecessary indeed.

Reply:
In fact @shargon, composition EntryOnly + CustomHash may be heavily used... it's a simple Enum attribution for us, and there's no other way to simulate this (as multiple scopes work as OR logic, not AND logic). This composition is AND logic. 
It would be nice to have some voices from DEX here, because I think this feature will help them A LOT.